centrallix-sysdoc/OSDriver_Authoring.md

```diff
@@ -115,497 +170,1032 @@ License:  Copyright (C) 2001-2011 LightSys Technology Services. See LICENSE.txt
+    - `obj->Pathname->OpenCtl : pStruct[]`: Parameters for the open() operation, as defined by the driver author. These are specified in the path in a similar way to URLs (`example.qy?param1=value&param2=other_value`).  Drivers typically only use `obj->Pathname->OpenCtl[obj->SubPtr]` (see SubPtr below) to retrieve their own parameters, ignoring parameters passed to other drivers in the path.
 
-As a part of the Open process, the OS driver will normally allocate an internal structure to represent the current open object, and will return that structure as a void* data type in the return value.  This pointer will be then passed to each of the other driver entry point functions, with the exception of QueryFetch, QueryDelete, and Query- Close, which will be discussed later.
+    - `obj->SubPtr : short`: The number of components in the path that are a part of the path to the root node object, including the `.` for the top level directory.  For example, in the above path of `/data/file.csv`, the path would be internally represented as `./ data/ file.csv`, so SubPtr is 3.
 
-The Open() routine is called with five parameters:
+    - `obj->SubCnt : short`: _The driver should set this value_ to show the number of components it controls.  This includes the root node object, so `SubCnt` will always be at least 1.  For example, when opening `/data/file.csv/rows/1`, the CSV driver will read the `SubPtr` of 3 (see above), representing `./ data/ file.csv`. It will then set a `SubCnt` of 3, representing that it will control `file.csv /rows /1`.  (The driver only sets `SubCnt`, `SubPtr` is provided.)
 
-- obj (pObject)
-    This is a pointer to the Object sturcture maintained by the OSML.  This structure will contain some important fields for processing the open() request.
+    - `obj->Prev : pObject`: The underlying object as opened by the next-lower-level driver. The file can be accessed and parsed by calling functions and passing this pointer to them (such as the st_parse functions, see below).  **DO NOT attempt to open the file directly with a call like `fopen()`,** as this would require hard coding the path to the root directory of the object system, which *will* break if the code runs on another machine.
 
-    obj->Mode is a bitmask of the O_* flags, which include O_RDONLY, O_WRONLY, O_RDWR, O_CREAT, O_TRUNC, and O_EXCL.
+    - `obj->Prev->Flags : short`: Contains some useful flags about the underlying object, such as:
+        - `OBJ_F_CREATED`: The underlying object was just created by this open() operation.  In that case, this driver is expected to create the node with `snNewNode()` (see later in this document) as long as `obj->Mode` contains `O_CREAT`.
+	<!-- TODO: Greg - Should we document any more of the flags? OBJ_F_ROOTNODE? OBJ_F_DELETE? OBJ_F_METAONLY? -->
```

Comment:
I don't think the other flags need to be documented here.  Some of them are for future use and some are for internal use.  Driver authors will eventually need to understand OBJ_F_UNMANAGED, but that's a later topic (objUnmanageObject) for drivers that need to open up additional objects that won't ever be directly visible to the OSML API consumer.